File: packages/tonic-core/src/weaveGit/mergeDriver.ts
/**
 * Keep in sync with `merge-tonic-lib/tonic/weave_git/merge_driver.py` — run Python + TS tests when editing.
 */

import * as fs from "node:fs";

import { currentLines, initialState, mergeStates } from "../core";
import { canonicalTextLines, normalizeLf, sha256HexBytes } from "./hashutil";
import { pathEntryForFile } from "./manifest";
import { stateHash } from "./replay";
import type { PathManifestEntry } from "./types";
import { weaveBlobPath } from "./verify";

export type MergeDriverResult = {
  exitCode: number;
  mergedText: string;
  manifestEntry: PathManifestEntry | null;
  stderr: string[];
  serializedWeave: string | null;
};

export type LoadStateFn = () => string | null;

export function mergeDriverExitCodeForAnnotated(annotatedLines: string[]): number {
  return annotatedLines.some((line) => line.includes("<<<<<<<")) ? 1 : 0;
}

export function mergeDriverCompatibilityErrors(params: {
  textBase: string;
  textOurs: string;
  textTheirs: string;
  sBase: string;
  sOurs: string;
  sTheirs: string;
  weaveFormatVersion: string;
  diffEngineId: string;
  manifestEntryBase?: PathManifestEntry | null;
  manifestEntryOurs?: PathManifestEntry | null;
  manifestEntryTheirs?: PathManifestEntry | null;
}): string[] {
  const errs: string[] = [];
  const triple: [string, string, string][] = [
    ["base", params.textBase, params.sBase],
    ["ours", params.textOurs, params.sOurs],
    ["theirs", params.textTheirs, params.sTheirs],
  ];
  for (const [label, text, state] of triple) {
    try {
      const wantLines = canonicalTextLines(text);
      const gotLines = currentLines(state);
      if (gotLines.length !== wantLines.length || gotLines.some((g, i) => g !== wantLines[i])) {
        errs.push(
          `${label}: weave current_lines do not match Git text (replay/text mismatch); ` +
            `refuse merge_states without aligned artifacts`,
        );
      }
    } catch (e) {
      errs.push(`${label}: current_lines failed: ${e instanceof Error ? e.message : String(e)}`);
    }
  }

  const entries = [params.manifestEntryBase, params.manifestEntryOurs, params.manifestEntryTheirs];
  if (entries.some((e) => e != null)) {
    for (const [label, state, entry] of [
      ["base", params.sBase, params.manifestEntryBase],
      ["ours", params.sOurs, params.manifestEntryOurs],
      ["theirs", params.sTheirs, params.manifestEntryTheirs],
    ] as const) {
      if (!entry) {
        continue;
      }
      const wss = entry.weave_serialized_sha;
      if (typeof wss === "string" && wss && stateHash(state) !== wss) {
        errs.push(`${label}: manifest weave_serialized_sha does not match loaded state`);
      }
      const ve = entry.weave_format_version;
      const de = entry.diff_engine_id;
      if (typeof ve === "string" && ve && ve !== params.weaveFormatVersion) {
        errs.push(
          `${label}: weave_format_version mismatch (manifest ${JSON.stringify(ve)} vs driver ${JSON.stringify(params.weaveFormatVersion)})`,
        );
      }
      if (typeof de === "string" && de && de !== params.diffEngineId) {
        errs.push(
          `${label}: diff_engine_id mismatch (manifest ${JSON.stringify(de)} vs driver ${JSON.stringify(params.diffEngineId)})`,
        );
      }
    }

    const po = params.manifestEntryOurs?.parent_weave_shas;
    const pt = params.manifestEntryTheirs?.parent_weave_shas;
    if (Array.isArray(po) && Array.isArray(pt) && po.length > 0 && pt.length > 0) {
      const bsha = stateHash(params.sBase);
      if (!po.includes(bsha)) {
        errs.push("ours: parent_weave_shas does not include merge-base weave_serialized_sha");
      }
      if (!pt.includes(bsha)) {
        errs.push("theirs: parent_weave_shas does not include merge-base weave_serialized_sha");
      }
    }
  }

  return errs;
}

export function runOptionAMerge(params: {
  textBase: string;
  textOurs: string;
  textTheirs: string;
  loadStateOurs: LoadStateFn;
  loadStateTheirs: LoadStateFn;
  loadStateBase: LoadStateFn;
  weaveFormatVersion: string;
  diffEngineId: string;
  strict?: boolean;
  manifestEntryBase?: PathManifestEntry | null;
  manifestEntryOurs?: PathManifestEntry | null;
  manifestEntryTheirs?: PathManifestEntry | null;
}): MergeDriverResult {
  const strict = params.strict !== false;
  const stderr: string[] = [];
  const sOurs = params.loadStateOurs();
  const sTheirs = params.loadStateTheirs();
  const sBase = params.loadStateBase();
  if (sOurs === null || sTheirs === null) {
    stderr.push("missing ours/theirs weave state");
    return { exitCode: 1, mergedText: "", manifestEntry: null, stderr, serializedWeave: null };
  }
  if (sBase === null) {
    const msg =
      "TONIC_WEAVE_DEGRADED: merge-base weave missing; snapshot merge only. " +
      "Strict repos should reject this commit.";
    stderr.push(msg);
    if (strict) {
      return { exitCode: 1, mergedText: "", manifestEntry: null, stderr, serializedWeave: null };
    }
    const left = normalizeLf(params.textOurs).split("\n");
    const right = normalizeLf(params.textTheirs).split("\n");
    if (left.length && left[left.length - 1] === "") {
      left.pop();
    }
    if (right.length && right[right.length - 1] === "") {
      right.pop();
    }
    const [mergedState, ann] = mergeStates(initialState(left), initialState(right));
    const text = ann.join("\n") + (ann.length ? "\n" : "");
    const [, entry] = pathEntryForFile({
      relPath: ".",
      textCanonical: text,
      serializedWeave: mergedState,
      weaveFormatVersion: params.weaveFormatVersion,
      diffEngineId: params.diffEngineId,
      degraded: true,
    });
    return {
      exitCode: mergeDriverExitCodeForAnnotated(ann),
      mergedText: text,
      manifestEntry: entry,
      stderr,
      serializedWeave: mergedState,
    };
  }

  const compat = mergeDriverCompatibilityErrors({
    textBase: params.textBase,
    textOurs: params.textOurs,
    textTheirs: params.textTheirs,
    sBase,
    sOurs,
    sTheirs,
    weaveFormatVersion: params.weaveFormatVersion,
    diffEngineId: params.diffEngineId,
    manifestEntryBase: params.manifestEntryBase,
    manifestEntryOurs: params.manifestEntryOurs,
    manifestEntryTheirs: params.manifestEntryTheirs,
  });
  for (const line of compat) {
    stderr.push(line);
  }
  if (compat.length > 0) {
    stderr.push("TONIC_WEAVE_MERGE_INCOMPAT: replay/manifest compatibility check failed");
    if (strict) {
      return { exitCode: 1, mergedText: "", manifestEntry: null, stderr, serializedWeave: null };
    }
    const left = normalizeLf(params.textOurs).split("\n");
    const right = normalizeLf(params.textTheirs).split("\n");
    if (left.length && left[left.length - 1] === "") {
      left.pop();
    }
    if (right.length && right[right.length - 1] === "") {
      right.pop();
    }
    const [mergedState, ann] = mergeStates(initialState(left), initialState(right));
    const text = ann.join("\n") + (ann.length ? "\n" : "");
    const [, entry] = pathEntryForFile({
      relPath: ".",
      textCanonical: text,
      serializedWeave: mergedState,
      weaveFormatVersion: params.weaveFormatVersion,
      diffEngineId: params.diffEngineId,
      degraded: true,
    });
    return {
      exitCode: mergeDriverExitCodeForAnnotated(ann),
      mergedText: text,
      manifestEntry: entry,
      stderr,
      serializedWeave: mergedState,
    };
  }

  try {
    const [mergedState, annotated] = mergeStates(sOurs, sTheirs);
    const linesOut = annotated;
    const text = linesOut.join("\n") + (linesOut.length ? "\n" : "");
    const [, entry] = pathEntryForFile({
      relPath: ".",
      textCanonical: text,
      serializedWeave: mergedState,
      weaveFormatVersion: params.weaveFormatVersion,
      diffEngineId: params.diffEngineId,
      degraded: false,
    });
    return {
      exitCode: mergeDriverExitCodeForAnnotated(linesOut),
      mergedText: text,
      manifestEntry: entry,
      stderr,
      serializedWeave: mergedState,
    };
  } catch (e) {
    stderr.push(`merge_states failed: ${e instanceof Error ? e.message : String(e)}`);
    return { exitCode: 1, mergedText: "", manifestEntry: null, stderr, serializedWeave: null };
  }
}

export function loadStateFromWeaveRoot(weaveRoot: string, weaveSerializedSha: string): string | null {
  const p = weaveBlobPath(weaveRoot, weaveSerializedSha);
  if (!fs.existsSync(p) || !fs.statSync(p).isFile()) {
    return null;
  }
  const data = fs.readFileSync(p);
  if (sha256HexBytes(data) !== weaveSerializedSha) {
    return null;
  }
  return data.toString("utf8");
}

File: merge-tonic-lib/tonic/weave_git/merge_driver.py
"""Git merge driver Option A: `merge_states` on full weave states with merge-base check."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable, Mapping

from tonic import current_lines, initial_state, merge_states

from .hashutil import canonical_text_lines, normalize_lf, sha256_hex_bytes
from .manifest import path_entry_for_file
from .replay import state_hash
from .types import PathManifestEntry
from .verify import weave_blob_path


@dataclass
class MergeDriverResult:
    exit_code: int
    merged_text: str
    manifest_entry: PathManifestEntry | None
    stderr: list[str]
    serialized_weave: str | None = None


LoadStateFn = Callable[[], str | None]


def merge_driver_exit_code_for_annotated(annotated_lines: list[str]) -> int:
    """Git merge drivers return 1 when the merged file still contains conflict markers."""
    return 1 if any("<<<<<<<" in line for line in annotated_lines) else 0


def merge_driver_compatibility_errors(
    *,
    text_base: str,
    text_ours: str,
    text_theirs: str,
    s_base: str,
    s_ours: str,
    s_theirs: str,
    weave_format_version: str,
    diff_engine_id: str,
    manifest_entry_base: Mapping[str, Any] | None = None,
    manifest_entry_ours: Mapping[str, Any] | None = None,
    manifest_entry_theirs: Mapping[str, Any] | None = None,
) -> list[str]:
    """§5.4 step 4: text ↔ weave alignment, manifest hashes, engine/version, optional parent weave chain."""
    errs: list[str] = []
    triple = [
        ("base", text_base, s_base),
        ("ours", text_ours, s_ours),
        ("theirs", text_theirs, s_theirs),
    ]
    for label, text, state in triple:
        try:
            want_lines = canonical_text_lines(text)
            got_lines = current_lines(state)
        except Exception as e:  # noqa: BLE001
            errs.append(f"{label}: current_lines failed: {e}")
            continue
        if got_lines != want_lines:
            errs.append(
                f"{label}: weave current_lines do not match Git text (replay/text mismatch); "
                f"refuse merge_states without aligned artifacts"
            )

    entries = (manifest_entry_base, manifest_entry_ours, manifest_entry_theirs)
    if any(e is not None for e in entries):
        for label, state, entry in (
            ("base", s_base, manifest_entry_base),
            ("ours", s_ours, manifest_entry_ours),
            ("theirs", s_theirs, manifest_entry_theirs),
        ):
            if entry is None:
                continue
            wss = entry.get("weave_serialized_sha")
            if isinstance(wss, str) and wss and state_hash(state) != wss:
                errs.append(f"{label}: manifest weave_serialized_sha does not match loaded state")
            ve = entry.get("weave_format_version")
            de = entry.get("diff_engine_id")
            if isinstance(ve, str) and ve and ve != weave_format_version:
                errs.append(f"{label}: weave_format_version mismatch (manifest {ve!r} vs driver {weave_format_version!r})")
            if isinstance(de, str) and de and de != diff_engine_id:
                errs.append(f"{label}: diff_engine_id mismatch (manifest {de!r} vs driver {diff_engine_id!r})")

        po = manifest_entry_ours.get("parent_weave_shas") if manifest_entry_ours else None
        pt = manifest_entry_theirs.get("parent_weave_shas") if manifest_entry_theirs else None
        if (
            isinstance(po, list)
            and isinstance(pt, list)
            and len(po) > 0
            and len(pt) > 0
        ):
            bsha = state_hash(s_base)
            if bsha not in po:
                errs.append("ours: parent_weave_shas does not include merge-base weave_serialized_sha")
            if bsha not in pt:
                errs.append("theirs: parent_weave_shas does not include merge-base weave_serialized_sha")

    return errs


def run_option_a_merge(
    *,
    text_base: str,
    text_ours: str,
    text_theirs: str,
    load_state_ours: LoadStateFn,
    load_state_theirs: LoadStateFn,
    load_state_base: LoadStateFn,
    weave_format_version: str,
    diff_engine_id: str,
    strict: bool = True,
    manifest_entry_base: Mapping[str, Any] | None = None,
    manifest_entry_ours: Mapping[str, Any] | None = None,
    manifest_entry_theirs: Mapping[str, Any] | None = None,
) -> MergeDriverResult:
    stderr: list[str] = []
    s_ours = load_state_ours()
    s_theirs = load_state_theirs()
    s_base = load_state_base()
    if s_ours is None or s_theirs is None:
        msg = "missing ours/theirs weave state"
        stderr.append(msg)
        return MergeDriverResult(1, "", None, stderr, None)
    if s_base is None:
        msg = (
            "TONIC_WEAVE_DEGRADED: merge-base weave missing; snapshot merge only. "
            "Strict repos should reject this commit."
        )
        stderr.append(msg)
        if strict:
            return MergeDriverResult(1, "", None, stderr, None)
        left = normalize_lf(text_ours).splitlines()
        right = normalize_lf(text_theirs).splitlines()
        merged_state, ann = merge_states(initial_state(left), initial_state(right))
        text = "\n".join(ann) + ("\n" if ann else "")
        _, entry = path_entry_for_file(
            rel_path=".",
            text_canonical=text,
            serialized_weave=merged_state,
            weave_format_version=weave_format_version,
            diff_engine_id=diff_engine_id,
            degraded=True,
        )
        return MergeDriverResult(
            merge_driver_exit_code_for_annotated(ann), text, entry, stderr, merged_state
        )

    compat = merge_driver_compatibility_errors(
        text_base=text_base,
        text_ours=text_ours,
        text_theirs=text_theirs,
        s_base=s_base,
        s_ours=s_ours,
        s_theirs=s_theirs,
        weave_format_version=weave_format_version,
        diff_engine_id=diff_engine_id,
        manifest_entry_base=manifest_entry_base,
        manifest_entry_ours=manifest_entry_ours,
        manifest_entry_theirs=manifest_entry_theirs,
    )
    for line in compat:
        stderr.append(line)
    if compat:
        msg = "TONIC_WEAVE_MERGE_INCOMPAT: replay/manifest compatibility check failed"
        stderr.append(msg)
        if strict:
            return MergeDriverResult(1, "", None, stderr, None)
        left = normalize_lf(text_ours).splitlines()
        right = normalize_lf(text_theirs).splitlines()
        merged_state, ann = merge_states(initial_state(left), initial_state(right))
        text = "\n".join(ann) + ("\n" if ann else "")
        _, entry = path_entry_for_file(
            rel_path=".",
            text_canonical=text,
            serialized_weave=merged_state,
            weave_format_version=weave_format_version,
            diff_engine_id=diff_engine_id,
            degraded=True,
        )
        return MergeDriverResult(
            merge_driver_exit_code_for_annotated(ann), text, entry, stderr, merged_state
        )

    try:
        merged_state, annotated = merge_states(s_ours, s_theirs)
    except Exception as e:  # noqa: BLE001
        stderr.append(f"merge_states failed: {e}")
        return MergeDriverResult(1, "", None, stderr, None)
    lines_out = annotated
    text = "\n".join(lines_out) + ("\n" if lines_out else "")
    _, entry = path_entry_for_file(
        rel_path=".",
        text_canonical=text,
        serialized_weave=merged_state,
        weave_format_version=weave_format_version,
        diff_engine_id=diff_engine_id,
        degraded=False,
    )
    return MergeDriverResult(
        merge_driver_exit_code_for_annotated(lines_out), text, entry, stderr, merged_state
    )


def load_state_from_weave_root(weave_root: Path, weave_serialized_sha: str) -> str | None:
    p = weave_blob_path(weave_root, weave_serialized_sha)
    if not p.is_file():
        return None
    data = p.read_bytes()
    if sha256_hex_bytes(data) != weave_serialized_sha:
        return None
    return data.decode("utf-8")
